### PR TITLE
Create first BLE-only board, Micro:Bit v2

### LINKED ISSUE #4546: Support for micro:bit v2 (nRF52833)
Following the guide at https://learn.adafruit.com/how-to-add-a-new-board-to-circuitpython/ :

* copied `boards/simmel`, which is another nRF52833-based board;
* updated `mpconfigboard.mk` with
```
USB_VID = 0x0d28
USB_PID = 0x0204
USB_PRODUCT = "micro:bit"
USB_MANUFACTURER = "BBC"
```
* updated `mpconfigboard.h` with
```
#define MICROPY_HW_BOARD_NAME       "micro:bit"

#define DEFAULT_I2C_BUS_SCL         (&pin_P0_26)
#define DEFAULT_I2C_BUS_SDA         (&pin_P1_00)

#define DEFAULT_UART_BUS_RX     (&pin_P0_06)
#define DEFAULT_UART_BUS_TX     (&pin_P1_08)
```

According to the guide, this should have been enough to get the REPL running over the serial interface. The connection to the serial interface succeeds, but in response to any input, I just see the amber status LED flashing, but no output.

What am I missing?

The pinmap (and other board data) are published at https://tech.microbit.org/hardware/schematic/

It would seem that CircuitPython had supported the original (nRF51822) micro:bit until 2018 when the support for all nRF51 boards was dropped (05bd5d4; #591)

### LINKED ISSUE #4680: Clean up nRF sleep code a bit more
Most notably, let's address this. https://github.com/adafruit/circuitpython/pull/4236#pullrequestreview-615739225

We are merging with this outstanding to unblock some other PR's.

### PR DESCRIPTION
**Pull request by tannewt**
This fixes build issues with USB off, tweaks the README to allow
for BLE-only boards and adds the Micro:Bit v2 definition.

Fixes #4546 Fixes #4680

### CODE REVIEW

**Review by jepler — APPROVED** [review]
The debug printing changes looked weird, but if it's intended then it's OK to merge.  No testing performed.

File: ports/nrf/common-hal/alarm/SleepMemory.c [hunk 1]
```diff
@@ -92,7 +84,7 @@ void alarm_sleep_memory_reset(void) {
     if (!is_sleep_memory_valid()) {
         initialize_sleep_memory();
         #ifdef NRF_DEBUG_PRINT
-        dbg_printf("sleep memory initialized\r\n");
+        mp_printf(&mp_plat_print, "sleep memory initialized\r\n");
```
[the remark this answers is not in the document]
Comment (tannewt) [reply ports/nrf/common-hal/alarm/SleepMemory.c:87]: Yup, this was intentional and should be guarded by an `#if` still.

**Comment by tyomitch** [comment]
@tannewt I see that the `Makefile` now relies on a script `hexmerge.py` which is not in the source tree.
Where is it supposed to be taken from?

**Comment by jerryneedell** [comment]
hexmerge.py is part of the intelhex package.
I installed it via `pip3 install intelhex`

**Comment by jerryneedell** [comment]
I would be happy to help with testing this build, but I am completely baffled about how to use the board.
I have occasionally managed to get the serial device to spit out the CP version info (see screenshot), but nothing else. I can see the CIRCPY BLE device but have made no progress connecting to it.
Should I be able to get to the REPL via something like screen? I can connect, but get no responses....
Clearly, I am missing something. Are there any instructions for using code.circuitpython.org?

Screen Shot 2021-07-16 at 7 19 17 AM

**Comment by tannewt** [comment]
> @tannewt I see that the `Makefile` now relies on a script `hexmerge.py` which is not in the source tree.
> Where is it supposed to be taken from?

We store python build dependencies in `requirements-dev.txt` so you should be able to `pip install -r requirements-dev.txt` to get everything you need.

**Comment by tannewt** [comment]
> I would be happy to help with testing this build, but I am completely baffled about how to use the board.
> I have occasionally managed to get the serial device to spit out the CP version info (see screenshot), but nothing else. I can see the CIRCPY BLE device but have made no progress connecting to it.
> Should I be able to get to the REPL via something like screen? I can connect, but get no responses....
> Clearly, I am missing something. Are there any instructions for using code.circuitpython.org?
> 
> Screen Shot 2021-07-16 at 7 19 17 AM

Thanks for trying this out! I know its super early for apps on the host side.

The USB serial is a conversion of the UART debug output. I'm going to turn this off shortly to make it clearer that this is only meant for BLE use. (Serial is really nice for debugging though.)

The most instructions I've written are on the page itself. I should make it clearer that BLE stuff will only work with Mac Chrome and Android Chrome. Chrome OS and Windows don't support secure characteristics yet (the chrome folks are working on it now I think.)

I think there are some crashes on the device side too unfortunately. I was trying to debug it but ran into debugger and binary blob issues.

So, don't spend too much time hitting a wall with this. It's still very early and will get better as the host side matures.

**Comment by jerryneedell** [comment]
I am using an M1 Mac -- I can try going back to an Intel Mac as well if that may be a problem.

**Comment by spoelstraethan** [comment]
@tannewt would there be many changes for the micro:bit v1 to run CircuitPython or is the slightly lower spec hardware not capable?

This is a good comparison of both the new features and the processor/memory differences.

https://kitronik.co.uk/blogs/resources/explore-micro-bit-v1-microbit-v2-differences

**Comment by tyomitch** [comment]
@spoelstraethan I doubt CircuitPython can be squeezed into 16 KB RAM; even 32 KB boards are a very tight fit.

**Comment by tannewt** [comment]
> @tannewt would there be many changes for the micro:bit v1 to run CircuitPython or is the slightly lower spec hardware not capable?

No, I don't plan on supporting the v1. As tyomitch points out, the change from 128k ram to 16k ram is a huge downgrade. We regularly have folks frustrated with the 32k ram on the SAMD21.

**Pull request merged**